{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "16S.ipynb",
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/wutto/tutorialAmpliconSequencing/blob/main/16S.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**A. à¸à¸²à¸£à¸¥à¸‡à¹‚à¸›à¸£à¹à¸à¸£à¸¡ mamba environment à¸ªà¸³à¸«à¸£à¸±à¸šà¸à¸²à¸£à¸¥à¸‡à¹‚à¸›à¸£à¹à¸à¸£à¸¡ QIIME2**\n",
        "\n"
      ],
      "metadata": {
        "id": "CqTIwbSxALvV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget \"https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh\"\n",
        "\n",
        "!bash Miniforge3-$(uname)-$(uname -m).sh\n",
        "\n",
        "!export PATH=$PATH:/root/miniforge3/bin/mamba"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CtfgXcAjAJ8w",
        "outputId": "5c91755e-7e5b-47ad-98bf-293e8b68a1d2",
        "collapsed": true
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-05-22 07:52:17--  https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh\n",
            "Resolving github.com (github.com)... 140.82.116.4\n",
            "Connecting to github.com (github.com)|140.82.116.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github.com/conda-forge/miniforge/releases/download/25.3.0-3/Miniforge3-Linux-x86_64.sh [following]\n",
            "--2025-05-22 07:52:17--  https://github.com/conda-forge/miniforge/releases/download/25.3.0-3/Miniforge3-Linux-x86_64.sh\n",
            "Reusing existing connection to github.com:443.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/221584272/11d9f638-f965-42b5-be9a-b27df154f8b0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250522%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250522T075217Z&X-Amz-Expires=300&X-Amz-Signature=0708fd0a81908a6c9d31e2004bbac16d92b6e64474f78278102393509c039752&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DMiniforge3-Linux-x86_64.sh&response-content-type=application%2Foctet-stream [following]\n",
            "--2025-05-22 07:52:17--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/221584272/11d9f638-f965-42b5-be9a-b27df154f8b0?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20250522%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20250522T075217Z&X-Amz-Expires=300&X-Amz-Signature=0708fd0a81908a6c9d31e2004bbac16d92b6e64474f78278102393509c039752&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DMiniforge3-Linux-x86_64.sh&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 93676241 (89M) [application/octet-stream]\n",
            "Saving to: â€˜Miniforge3-Linux-x86_64.shâ€™\n",
            "\n",
            "Miniforge3-Linux-x8 100%[===================>]  89.34M   149MB/s    in 0.6s    \n",
            "\n",
            "2025-05-22 07:52:18 (149 MB/s) - â€˜Miniforge3-Linux-x86_64.shâ€™ saved [93676241/93676241]\n",
            "\n",
            "\n",
            "Welcome to Miniforge3 25.3.0-3\n",
            "\n",
            "In order to continue the installation process, please review the license\n",
            "agreement.\n",
            "Please, press ENTER to continue\n",
            ">>> \n",
            "Miniforge installer code uses BSD-3-Clause license as stated below.\n",
            "\n",
            "Binary packages that come with it have their own licensing terms\n",
            "and by installing miniforge you agree to the licensing terms of individual\n",
            "packages as well. They include different OSI-approved licenses including\n",
            "the GNU General Public License and can be found in pkgs/<pkg-name>/info/licenses\n",
            "folders.\n",
            "\n",
            "Miniforge installer comes with a bootstrapping executable that is used\n",
            "when installing miniforge and is deleted after miniforge is installed.\n",
            "The bootstrapping executable uses micromamba, cli11, cpp-filesystem,\n",
            "curl, c-ares, krb5, libarchive, libev, lz4, nghttp2, openssl, libsolv,\n",
            "nlohmann-json, reproc and zstd which are licensed under BSD-3-Clause,\n",
            "MIT and OpenSSL licenses. Licenses and copyright notices of these\n",
            "projects can be found at the following URL.\n",
            "https://github.com/conda-forge/micromamba-feedstock/tree/master/recipe.\n",
            "\n",
            "=============================================================================\n",
            "\n",
            "Copyright (c) 2019-2022, conda-forge\n",
            "All rights reserved.\n",
            "\n",
            "Redistribution and use in source and binary forms, with or without\n",
            "modification, are permitted provided that the following conditions are met:\n",
            "\u001b[K\n",
            "1. Redistributions of source code must retain the above copyright notice, this\n",
            "list of conditions and the following disclaimer.\n",
            "\u001b[K\n",
            "2. Redistributions in binary form must reproduce the above copyright notice,\n",
            "this list of conditions and the following disclaimer in the documentation\n",
            "and/or other materials provided with the distribution.\n",
            "\u001b[K\n",
            "3. Neither the name of the copyright holder nor the names of its contributors\n",
            "may be used to endorse or promote products derived from this software without\n",
            "specific prior written permission.\n",
            "\u001b[K\n",
            "THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND\n",
            "ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED\n",
            "WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE\n",
            "DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE\n",
            "FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\n",
            "DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\n",
            "SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n",
            "CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\n",
            "OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\n",
            "OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n",
            "\u001b[K\n",
            "\n",
            "\n",
            "Do you accept the license terms? [yes|no]\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "\n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "\n",
            "\n",
            "Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> Please answer 'yes' or 'no':'\n",
            ">>> \n",
            "Please answer 'yes' or 'no':'\n",
            ">>> yes\n",
            "\n",
            "Miniforge3 will now be installed into this location:\n",
            "/root/miniforge3\n",
            "\n",
            "  - Press ENTER to confirm the location\n",
            "  - Press CTRL-C to abort the installation\n",
            "  - Or specify a different location below\n",
            "    Note: environment variables will NOT be expanded.\n",
            "\n",
            "[/root/miniforge3] >>> \n",
            "PREFIX=/root/miniforge3\n",
            "Unpacking payload ...\n",
            "Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2\n",
            "Extracting ca-certificates-2025.4.26-hbd8a1cb_0.conda\n",
            "Extracting ld_impl_linux-64-2.43-h712a8e2_4.conda\n",
            "Extracting libgomp-15.1.0-h767d61c_2.conda\n",
            "Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2\n",
            "Extracting python_abi-3.12-7_cp312.conda\n",
            "Extracting tzdata-2025b-h78e105d_0.conda\n",
            "Extracting _openmp_mutex-4.5-2_gnu.tar.bz2\n",
            "Extracting libgcc-15.1.0-h767d61c_2.conda\n",
            "Extracting c-ares-1.34.5-hb9d3cd8_0.conda\n",
            "Extracting libexpat-2.7.0-h5888daf_0.conda\n",
            "Extracting libffi-3.4.6-h2dba641_1.conda\n",
            "Extracting libgcc-ng-15.1.0-h69a702a_2.conda\n",
            "Extracting libiconv-1.18-h4ce23a2_1.conda\n",
            "Extracting liblzma-5.8.1-hb9d3cd8_1.conda\n",
            "Extracting libstdcxx-15.1.0-h8f9b012_2.conda\n",
            "Extracting libzlib-1.3.1-hb9d3cd8_2.conda\n",
            "Extracting ncurses-6.5-h2d0b736_3.conda\n",
            "Extracting openssl-3.5.0-h7b32b05_1.conda\n",
            "Extracting reproc-14.2.5.post0-hb9d3cd8_0.conda\n",
            "Extracting bzip2-1.0.8-h4bc722e_7.conda\n",
            "Extracting cpp-expected-1.1.0-hff21bea_1.conda\n",
            "Extracting fmt-11.1.4-h07f6e7f_1.conda\n",
            "Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2\n",
            "Extracting libedit-3.1.20250104-pl5321h7949ede_0.conda\n",
            "Extracting libev-4.33-hd590300_2.conda\n",
            "Extracting libnsl-2.0.1-hd590300_0.conda\n",
            "Extracting libsolv-0.7.32-h7955e40_2.conda\n",
            "Extracting libsqlite-3.49.2-hee588c1_0.conda\n",
            "Extracting libssh2-1.11.1-hcf80075_0.conda\n",
            "Extracting libstdcxx-ng-15.1.0-h4852527_2.conda\n",
            "Extracting libuuid-2.38.1-h0b41bf4_0.conda\n",
            "Extracting libxcrypt-4.4.36-hd590300_1.conda\n",
            "Extracting lz4-c-1.10.0-h5888daf_1.conda\n",
            "Extracting lzo-2.10-hd590300_1001.conda\n",
            "Extracting readline-8.2-h8c095d6_2.conda\n",
            "Extracting reproc-cpp-14.2.5.post0-h5888daf_0.conda\n",
            "Extracting simdjson-3.12.3-h84d6215_0.conda\n",
            "Extracting tk-8.6.13-noxft_h4845f30_101.conda\n",
            "Extracting yaml-cpp-0.8.0-h3f2d84a_0.conda\n",
            "Extracting zstd-1.5.7-hb8e6e7a_2.conda\n",
            "Extracting icu-75.1-he02047a_0.conda\n",
            "Extracting krb5-1.21.3-h659f571_0.conda\n",
            "Extracting libnghttp2-1.64.0-h161d5f1_0.conda\n",
            "Extracting nlohmann_json-3.11.3-he02047a_1.conda\n",
            "Extracting python-3.12.10-h9e4cc4f_0_cpython.conda\n",
            "Extracting libcurl-8.13.0-h332b0f4_0.conda\n",
            "Extracting libxml2-2.13.8-h4bc477f_0.conda\n",
            "Extracting menuinst-2.2.0-py312h7900ff3_0.conda\n",
            "Extracting archspec-0.2.5-pyhd8ed1ab_0.conda\n",
            "Extracting boltons-24.0.0-pyhd8ed1ab_1.conda\n",
            "Extracting brotli-python-1.1.0-py312h2ec8cdc_2.conda\n",
            "Extracting certifi-2025.4.26-pyhd8ed1ab_0.conda\n",
            "Extracting charset-normalizer-3.4.2-pyhd8ed1ab_0.conda\n",
            "Extracting colorama-0.4.6-pyhd8ed1ab_1.conda\n",
            "Extracting distro-1.9.0-pyhd8ed1ab_1.conda\n",
            "Extracting frozendict-2.4.6-py312h66e93f0_0.conda\n",
            "Extracting hpack-4.1.0-pyhd8ed1ab_0.conda\n",
            "Extracting hyperframe-6.1.0-pyhd8ed1ab_0.conda\n",
            "Extracting idna-3.10-pyhd8ed1ab_1.conda\n",
            "Extracting jsonpointer-3.0.0-py312h7900ff3_1.conda\n",
            "Extracting libarchive-3.7.7-h75ea233_4.conda\n",
            "Extracting packaging-25.0-pyh29332c3_1.conda\n",
            "Extracting platformdirs-4.3.8-pyhe01879c_0.conda\n",
            "Extracting pluggy-1.5.0-pyhd8ed1ab_1.conda\n",
            "Extracting pycosat-0.6.6-py312h66e93f0_2.conda\n",
            "Extracting pycparser-2.22-pyh29332c3_1.conda\n",
            "Extracting pysocks-1.7.1-pyha55dd90_7.conda\n",
            "Extracting ruamel.yaml.clib-0.2.8-py312h66e93f0_1.conda\n",
            "Extracting setuptools-80.1.0-pyhff2d567_0.conda\n",
            "Extracting truststore-0.10.1-pyh29332c3_0.conda\n",
            "Extracting wheel-0.45.1-pyhd8ed1ab_1.conda\n",
            "Extracting cffi-1.17.1-py312h06ac9bb_0.conda\n",
            "Extracting h2-4.2.0-pyhd8ed1ab_0.conda\n",
            "Extracting jsonpatch-1.33-pyhd8ed1ab_1.conda\n",
            "Extracting libmamba-2.1.1-h430c389_0.conda\n",
            "Extracting pip-25.1.1-pyh8b19718_0.conda\n",
            "Extracting ruamel.yaml-0.18.10-py312h66e93f0_0.conda\n",
            "Extracting tqdm-4.67.1-pyhd8ed1ab_1.conda\n",
            "Extracting libmambapy-2.1.1-py312h07448e0_0.conda\n",
            "Extracting mamba-2.1.1-had4a41a_0.conda\n",
            "Extracting zstandard-0.23.0-py312h66e93f0_2.conda\n",
            "Extracting conda-package-streaming-0.11.0-pyhd8ed1ab_1.conda\n",
            "Extracting urllib3-2.4.0-pyhd8ed1ab_0.conda\n",
            "Extracting requests-2.32.3-pyhd8ed1ab_1.conda\n",
            "Extracting conda-package-handling-2.4.0-pyh7900ff3_2.conda\n",
            "Extracting conda-25.3.0-py312h7900ff3_0.conda\n",
            "Extracting conda-libmamba-solver-25.3.0-pyhd8ed1ab_0.conda\n",
            "\n",
            "Installing base environment...\n",
            "\n",
            "Transaction\n",
            "\n",
            "  Prefix: /root/miniforge3\n",
            "\n",
            "  Updating specs:\n",
            "\n",
            "   - _libgcc_mutex==0.1=conda_forge\n",
            "   - ca-certificates==2025.4.26=hbd8a1cb_0\n",
            "   - ld_impl_linux-64==2.43=h712a8e2_4\n",
            "   - libgomp==15.1.0=h767d61c_2\n",
            "   - pybind11-abi==4=hd8ed1ab_3\n",
            "   - python_abi==3.12=7_cp312\n",
            "   - tzdata==2025b=h78e105d_0\n",
            "   - _openmp_mutex==4.5=2_gnu\n",
            "   - libgcc==15.1.0=h767d61c_2\n",
            "   - c-ares==1.34.5=hb9d3cd8_0\n",
            "   - libexpat==2.7.0=h5888daf_0\n",
            "   - libffi==3.4.6=h2dba641_1\n",
            "   - libgcc-ng==15.1.0=h69a702a_2\n",
            "   - libiconv==1.18=h4ce23a2_1\n",
            "   - liblzma==5.8.1=hb9d3cd8_1\n",
            "   - libstdcxx==15.1.0=h8f9b012_2\n",
            "   - libzlib==1.3.1=hb9d3cd8_2\n",
            "   - ncurses==6.5=h2d0b736_3\n",
            "   - openssl==3.5.0=h7b32b05_1\n",
            "   - reproc==14.2.5.0post0=hb9d3cd8_0\n",
            "   - bzip2==1.0.8=h4bc722e_7\n",
            "   - cpp-expected==1.1.0=hff21bea_1\n",
            "   - fmt==11.1.4=h07f6e7f_1\n",
            "   - keyutils==1.6.1=h166bdaf_0\n",
            "   - libedit==3.1.20250104=pl5321h7949ede_0\n",
            "   - libev==4.33=hd590300_2\n",
            "   - libnsl==2.0.1=hd590300_0\n",
            "   - libsolv==0.7.32=h7955e40_2\n",
            "   - libsqlite==3.49.2=hee588c1_0\n",
            "   - libssh2==1.11.1=hcf80075_0\n",
            "   - libstdcxx-ng==15.1.0=h4852527_2\n",
            "   - libuuid==2.38.1=h0b41bf4_0\n",
            "   - libxcrypt==4.4.36=hd590300_1\n",
            "   - lz4-c==1.10.0=h5888daf_1\n",
            "   - lzo==2.10=hd590300_1001\n",
            "   - readline==8.2=h8c095d6_2\n",
            "   - reproc-cpp==14.2.5.0post0=h5888daf_0\n",
            "   - simdjson==3.12.3=h84d6215_0\n",
            "   - tk==8.6.13=noxft_h4845f30_101\n",
            "   - yaml-cpp==0.8.0=h3f2d84a_0\n",
            "   - zstd==1.5.7=hb8e6e7a_2\n",
            "   - icu==75.1=he02047a_0\n",
            "   - krb5==1.21.3=h659f571_0\n",
            "   - libnghttp2==1.64.0=h161d5f1_0\n",
            "   - nlohmann_json==3.11.3=he02047a_1\n",
            "   - python==3.12.10=h9e4cc4f_0_cpython\n",
            "   - libcurl==8.13.0=h332b0f4_0\n",
            "   - libxml2==2.13.8=h4bc477f_0\n",
            "   - menuinst==2.2.0=py312h7900ff3_0\n",
            "   - archspec==0.2.5=pyhd8ed1ab_0\n",
            "   - boltons==24.0.0=pyhd8ed1ab_1\n",
            "   - brotli-python==1.1.0=py312h2ec8cdc_2\n",
            "   - certifi==2025.4.26=pyhd8ed1ab_0\n",
            "   - charset-normalizer==3.4.2=pyhd8ed1ab_0\n",
            "   - colorama==0.4.6=pyhd8ed1ab_1\n",
            "   - distro==1.9.0=pyhd8ed1ab_1\n",
            "   - frozendict==2.4.6=py312h66e93f0_0\n",
            "   - hpack==4.1.0=pyhd8ed1ab_0\n",
            "   - hyperframe==6.1.0=pyhd8ed1ab_0\n",
            "   - idna==3.10=pyhd8ed1ab_1\n",
            "   - jsonpointer==3.0.0=py312h7900ff3_1\n",
            "   - libarchive==3.7.7=h75ea233_4\n",
            "   - packaging==25.0=pyh29332c3_1\n",
            "   - platformdirs==4.3.8=pyhe01879c_0\n",
            "   - pluggy==1.5.0=pyhd8ed1ab_1\n",
            "   - pycosat==0.6.6=py312h66e93f0_2\n",
            "   - pycparser==2.22=pyh29332c3_1\n",
            "   - pysocks==1.7.1=pyha55dd90_7\n",
            "   - ruamel.yaml.clib==0.2.8=py312h66e93f0_1\n",
            "   - setuptools==80.1.0=pyhff2d567_0\n",
            "   - truststore==0.10.1=pyh29332c3_0\n",
            "   - wheel==0.45.1=pyhd8ed1ab_1\n",
            "   - cffi==1.17.1=py312h06ac9bb_0\n",
            "   - h2==4.2.0=pyhd8ed1ab_0\n",
            "   - jsonpatch==1.33=pyhd8ed1ab_1\n",
            "   - libmamba==2.1.1=h430c389_0\n",
            "   - pip==25.1.1=pyh8b19718_0\n",
            "   - ruamel.yaml==0.18.10=py312h66e93f0_0\n",
            "   - tqdm==4.67.1=pyhd8ed1ab_1\n",
            "   - libmambapy==2.1.1=py312h07448e0_0\n",
            "   - mamba==2.1.1=had4a41a_0\n",
            "   - zstandard==0.23.0=py312h66e93f0_2\n",
            "   - conda-package-streaming==0.11.0=pyhd8ed1ab_1\n",
            "   - urllib3==2.4.0=pyhd8ed1ab_0\n",
            "   - requests==2.32.3=pyhd8ed1ab_1\n",
            "   - conda-package-handling==2.4.0=pyh7900ff3_2\n",
            "   - conda==25.3.0=py312h7900ff3_0\n",
            "   - conda-libmamba-solver==25.3.0=pyhd8ed1ab_0\n",
            "\n",
            "\n",
            "  Package                         Version  Build               Channel         Size\n",
            "â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
            "  Install:\n",
            "â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
            "\n",
            "  \u001b[32m+ _libgcc_mutex          \u001b[0m           0.1  conda_forge         conda-forge         \n",
            "  \u001b[32m+ _openmp_mutex          \u001b[0m           4.5  2_gnu               conda-forge         \n",
            "  \u001b[32m+ archspec               \u001b[0m         0.2.5  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ boltons                \u001b[0m        24.0.0  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ brotli-python          \u001b[0m         1.1.0  py312h2ec8cdc_2     conda-forge         \n",
            "  \u001b[32m+ bzip2                  \u001b[0m         1.0.8  h4bc722e_7          conda-forge         \n",
            "  \u001b[32m+ c-ares                 \u001b[0m        1.34.5  hb9d3cd8_0          conda-forge         \n",
            "  \u001b[32m+ ca-certificates        \u001b[0m     2025.4.26  hbd8a1cb_0          conda-forge         \n",
            "  \u001b[32m+ certifi                \u001b[0m     2025.4.26  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ cffi                   \u001b[0m        1.17.1  py312h06ac9bb_0     conda-forge         \n",
            "  \u001b[32m+ charset-normalizer     \u001b[0m         3.4.2  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ colorama               \u001b[0m         0.4.6  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ conda                  \u001b[0m        25.3.0  py312h7900ff3_0     conda-forge         \n",
            "  \u001b[32m+ conda-libmamba-solver  \u001b[0m        25.3.0  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ conda-package-handling \u001b[0m         2.4.0  pyh7900ff3_2        conda-forge         \n",
            "  \u001b[32m+ conda-package-streaming\u001b[0m        0.11.0  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ cpp-expected           \u001b[0m         1.1.0  hff21bea_1          conda-forge         \n",
            "  \u001b[32m+ distro                 \u001b[0m         1.9.0  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ fmt                    \u001b[0m        11.1.4  h07f6e7f_1          conda-forge         \n",
            "  \u001b[32m+ frozendict             \u001b[0m         2.4.6  py312h66e93f0_0     conda-forge         \n",
            "  \u001b[32m+ h2                     \u001b[0m         4.2.0  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ hpack                  \u001b[0m         4.1.0  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ hyperframe             \u001b[0m         6.1.0  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ icu                    \u001b[0m          75.1  he02047a_0          conda-forge         \n",
            "  \u001b[32m+ idna                   \u001b[0m          3.10  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ jsonpatch              \u001b[0m          1.33  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ jsonpointer            \u001b[0m         3.0.0  py312h7900ff3_1     conda-forge         \n",
            "  \u001b[32m+ keyutils               \u001b[0m         1.6.1  h166bdaf_0          conda-forge         \n",
            "  \u001b[32m+ krb5                   \u001b[0m        1.21.3  h659f571_0          conda-forge         \n",
            "  \u001b[32m+ ld_impl_linux-64       \u001b[0m          2.43  h712a8e2_4          conda-forge         \n",
            "  \u001b[32m+ libarchive             \u001b[0m         3.7.7  h75ea233_4          conda-forge         \n",
            "  \u001b[32m+ libcurl                \u001b[0m        8.13.0  h332b0f4_0          conda-forge         \n",
            "  \u001b[32m+ libedit                \u001b[0m  3.1.20250104  pl5321h7949ede_0    conda-forge         \n",
            "  \u001b[32m+ libev                  \u001b[0m          4.33  hd590300_2          conda-forge         \n",
            "  \u001b[32m+ libexpat               \u001b[0m         2.7.0  h5888daf_0          conda-forge         \n",
            "  \u001b[32m+ libffi                 \u001b[0m         3.4.6  h2dba641_1          conda-forge         \n",
            "  \u001b[32m+ libgcc                 \u001b[0m        15.1.0  h767d61c_2          conda-forge         \n",
            "  \u001b[32m+ libgcc-ng              \u001b[0m        15.1.0  h69a702a_2          conda-forge         \n",
            "  \u001b[32m+ libgomp                \u001b[0m        15.1.0  h767d61c_2          conda-forge         \n",
            "  \u001b[32m+ libiconv               \u001b[0m          1.18  h4ce23a2_1          conda-forge         \n",
            "  \u001b[32m+ liblzma                \u001b[0m         5.8.1  hb9d3cd8_1          conda-forge         \n",
            "  \u001b[32m+ libmamba               \u001b[0m         2.1.1  h430c389_0          conda-forge         \n",
            "  \u001b[32m+ libmambapy             \u001b[0m         2.1.1  py312h07448e0_0     conda-forge         \n",
            "  \u001b[32m+ libnghttp2             \u001b[0m        1.64.0  h161d5f1_0          conda-forge         \n",
            "  \u001b[32m+ libnsl                 \u001b[0m         2.0.1  hd590300_0          conda-forge         \n",
            "  \u001b[32m+ libsolv                \u001b[0m        0.7.32  h7955e40_2          conda-forge         \n",
            "  \u001b[32m+ libsqlite              \u001b[0m        3.49.2  hee588c1_0          conda-forge         \n",
            "  \u001b[32m+ libssh2                \u001b[0m        1.11.1  hcf80075_0          conda-forge         \n",
            "  \u001b[32m+ libstdcxx              \u001b[0m        15.1.0  h8f9b012_2          conda-forge         \n",
            "  \u001b[32m+ libstdcxx-ng           \u001b[0m        15.1.0  h4852527_2          conda-forge         \n",
            "  \u001b[32m+ libuuid                \u001b[0m        2.38.1  h0b41bf4_0          conda-forge         \n",
            "  \u001b[32m+ libxcrypt              \u001b[0m        4.4.36  hd590300_1          conda-forge         \n",
            "  \u001b[32m+ libxml2                \u001b[0m        2.13.8  h4bc477f_0          conda-forge         \n",
            "  \u001b[32m+ libzlib                \u001b[0m         1.3.1  hb9d3cd8_2          conda-forge         \n",
            "  \u001b[32m+ lz4-c                  \u001b[0m        1.10.0  h5888daf_1          conda-forge         \n",
            "  \u001b[32m+ lzo                    \u001b[0m          2.10  hd590300_1001       conda-forge         \n",
            "  \u001b[32m+ mamba                  \u001b[0m         2.1.1  had4a41a_0          conda-forge         \n",
            "  \u001b[32m+ menuinst               \u001b[0m         2.2.0  py312h7900ff3_0     conda-forge         \n",
            "  \u001b[32m+ ncurses                \u001b[0m           6.5  h2d0b736_3          conda-forge         \n",
            "  \u001b[32m+ nlohmann_json          \u001b[0m        3.11.3  he02047a_1          conda-forge         \n",
            "  \u001b[32m+ openssl                \u001b[0m         3.5.0  h7b32b05_1          conda-forge         \n",
            "  \u001b[32m+ packaging              \u001b[0m          25.0  pyh29332c3_1        conda-forge         \n",
            "  \u001b[32m+ pip                    \u001b[0m        25.1.1  pyh8b19718_0        conda-forge         \n",
            "  \u001b[32m+ platformdirs           \u001b[0m         4.3.8  pyhe01879c_0        conda-forge         \n",
            "  \u001b[32m+ pluggy                 \u001b[0m         1.5.0  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ pybind11-abi           \u001b[0m             4  hd8ed1ab_3          conda-forge         \n",
            "  \u001b[32m+ pycosat                \u001b[0m         0.6.6  py312h66e93f0_2     conda-forge         \n",
            "  \u001b[32m+ pycparser              \u001b[0m          2.22  pyh29332c3_1        conda-forge         \n",
            "  \u001b[32m+ pysocks                \u001b[0m         1.7.1  pyha55dd90_7        conda-forge         \n",
            "  \u001b[32m+ python                 \u001b[0m       3.12.10  h9e4cc4f_0_cpython  conda-forge         \n",
            "  \u001b[32m+ python_abi             \u001b[0m          3.12  7_cp312             conda-forge         \n",
            "  \u001b[32m+ readline               \u001b[0m           8.2  h8c095d6_2          conda-forge         \n",
            "  \u001b[32m+ reproc                 \u001b[0m  14.2.5.post0  hb9d3cd8_0          conda-forge         \n",
            "  \u001b[32m+ reproc-cpp             \u001b[0m  14.2.5.post0  h5888daf_0          conda-forge         \n",
            "  \u001b[32m+ requests               \u001b[0m        2.32.3  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ ruamel.yaml            \u001b[0m       0.18.10  py312h66e93f0_0     conda-forge         \n",
            "  \u001b[32m+ ruamel.yaml.clib       \u001b[0m         0.2.8  py312h66e93f0_1     conda-forge         \n",
            "  \u001b[32m+ setuptools             \u001b[0m        80.1.0  pyhff2d567_0        conda-forge         \n",
            "  \u001b[32m+ simdjson               \u001b[0m        3.12.3  h84d6215_0          conda-forge         \n",
            "  \u001b[32m+ tk                     \u001b[0m        8.6.13  noxft_h4845f30_101  conda-forge         \n",
            "  \u001b[32m+ tqdm                   \u001b[0m        4.67.1  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ truststore             \u001b[0m        0.10.1  pyh29332c3_0        conda-forge         \n",
            "  \u001b[32m+ tzdata                 \u001b[0m         2025b  h78e105d_0          conda-forge         \n",
            "  \u001b[32m+ urllib3                \u001b[0m         2.4.0  pyhd8ed1ab_0        conda-forge         \n",
            "  \u001b[32m+ wheel                  \u001b[0m        0.45.1  pyhd8ed1ab_1        conda-forge         \n",
            "  \u001b[32m+ yaml-cpp               \u001b[0m         0.8.0  h3f2d84a_0          conda-forge         \n",
            "  \u001b[32m+ zstandard              \u001b[0m        0.23.0  py312h66e93f0_2     conda-forge         \n",
            "  \u001b[32m+ zstd                   \u001b[0m         1.5.7  hb8e6e7a_2          conda-forge         \n",
            "\n",
            "  Summary:\n",
            "\n",
            "  Install: 88 packages\n",
            "\n",
            "  Total download: 0 B\n",
            "\n",
            "â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€\n",
            "\n",
            "\n",
            "\n",
            "Transaction starting\n",
            "\u001b[?25l\u001b[2K\u001b[0G\u001b[?25hLinking _libgcc_mutex-0.1-conda_forge\n",
            "Linking ca-certificates-2025.4.26-hbd8a1cb_0\n",
            "Linking ld_impl_linux-64-2.43-h712a8e2_4\n",
            "Linking libgomp-15.1.0-h767d61c_2\n",
            "Linking pybind11-abi-4-hd8ed1ab_3\n",
            "Linking python_abi-3.12-7_cp312\n",
            "Linking tzdata-2025b-h78e105d_0\n",
            "Linking _openmp_mutex-4.5-2_gnu\n",
            "Linking libgcc-15.1.0-h767d61c_2\n",
            "Linking c-ares-1.34.5-hb9d3cd8_0\n",
            "Linking libexpat-2.7.0-h5888daf_0\n",
            "Linking libffi-3.4.6-h2dba641_1\n",
            "Linking libgcc-ng-15.1.0-h69a702a_2\n",
            "Linking libiconv-1.18-h4ce23a2_1\n",
            "Linking liblzma-5.8.1-hb9d3cd8_1\n",
            "Linking libstdcxx-15.1.0-h8f9b012_2\n",
            "Linking libzlib-1.3.1-hb9d3cd8_2\n",
            "Linking ncurses-6.5-h2d0b736_3\n",
            "Linking openssl-3.5.0-h7b32b05_1\n",
            "Linking reproc-14.2.5.post0-hb9d3cd8_0\n",
            "Linking bzip2-1.0.8-h4bc722e_7\n",
            "Linking cpp-expected-1.1.0-hff21bea_1\n",
            "Linking fmt-11.1.4-h07f6e7f_1\n",
            "Linking keyutils-1.6.1-h166bdaf_0\n",
            "Linking libedit-3.1.20250104-pl5321h7949ede_0\n",
            "Linking libev-4.33-hd590300_2\n",
            "Linking libnsl-2.0.1-hd590300_0\n",
            "Linking libsolv-0.7.32-h7955e40_2\n",
            "Linking libsqlite-3.49.2-hee588c1_0\n",
            "Linking libssh2-1.11.1-hcf80075_0\n",
            "Linking libstdcxx-ng-15.1.0-h4852527_2\n",
            "Linking libuuid-2.38.1-h0b41bf4_0\n",
            "Linking libxcrypt-4.4.36-hd590300_1\n",
            "Linking lz4-c-1.10.0-h5888daf_1\n",
            "Linking lzo-2.10-hd590300_1001\n",
            "Linking readline-8.2-h8c095d6_2\n",
            "Linking reproc-cpp-14.2.5.post0-h5888daf_0\n",
            "Linking simdjson-3.12.3-h84d6215_0\n",
            "Linking tk-8.6.13-noxft_h4845f30_101\n",
            "Linking yaml-cpp-0.8.0-h3f2d84a_0\n",
            "Linking zstd-1.5.7-hb8e6e7a_2\n",
            "Linking icu-75.1-he02047a_0\n",
            "Linking krb5-1.21.3-h659f571_0\n",
            "Linking libnghttp2-1.64.0-h161d5f1_0\n",
            "Linking nlohmann_json-3.11.3-he02047a_1\n",
            "Linking python-3.12.10-h9e4cc4f_0_cpython\n",
            "Linking libcurl-8.13.0-h332b0f4_0\n",
            "Linking libxml2-2.13.8-h4bc477f_0\n",
            "Linking menuinst-2.2.0-py312h7900ff3_0\n",
            "Linking archspec-0.2.5-pyhd8ed1ab_0\n",
            "Linking boltons-24.0.0-pyhd8ed1ab_1\n",
            "Linking brotli-python-1.1.0-py312h2ec8cdc_2\n",
            "Linking certifi-2025.4.26-pyhd8ed1ab_0\n",
            "Linking charset-normalizer-3.4.2-pyhd8ed1ab_0\n",
            "Linking colorama-0.4.6-pyhd8ed1ab_1\n",
            "Linking distro-1.9.0-pyhd8ed1ab_1\n",
            "Linking frozendict-2.4.6-py312h66e93f0_0\n",
            "Linking hpack-4.1.0-pyhd8ed1ab_0\n",
            "Linking hyperframe-6.1.0-pyhd8ed1ab_0\n",
            "Linking idna-3.10-pyhd8ed1ab_1\n",
            "Linking jsonpointer-3.0.0-py312h7900ff3_1\n",
            "Linking libarchive-3.7.7-h75ea233_4\n",
            "Linking packaging-25.0-pyh29332c3_1\n",
            "Linking platformdirs-4.3.8-pyhe01879c_0\n",
            "Linking pluggy-1.5.0-pyhd8ed1ab_1\n",
            "Linking pycosat-0.6.6-py312h66e93f0_2\n",
            "Linking pycparser-2.22-pyh29332c3_1\n",
            "Linking pysocks-1.7.1-pyha55dd90_7\n",
            "Linking ruamel.yaml.clib-0.2.8-py312h66e93f0_1\n",
            "Linking setuptools-80.1.0-pyhff2d567_0\n",
            "Linking truststore-0.10.1-pyh29332c3_0\n",
            "Linking wheel-0.45.1-pyhd8ed1ab_1\n",
            "Linking cffi-1.17.1-py312h06ac9bb_0\n",
            "Linking h2-4.2.0-pyhd8ed1ab_0\n",
            "Linking jsonpatch-1.33-pyhd8ed1ab_1\n",
            "Linking libmamba-2.1.1-h430c389_0\n",
            "Linking pip-25.1.1-pyh8b19718_0\n",
            "Linking ruamel.yaml-0.18.10-py312h66e93f0_0\n",
            "Linking tqdm-4.67.1-pyhd8ed1ab_1\n",
            "Linking libmambapy-2.1.1-py312h07448e0_0\n",
            "Linking mamba-2.1.1-had4a41a_0\n",
            "Linking zstandard-0.23.0-py312h66e93f0_2\n",
            "Linking conda-package-streaming-0.11.0-pyhd8ed1ab_1\n",
            "Linking urllib3-2.4.0-pyhd8ed1ab_0\n",
            "Linking requests-2.32.3-pyhd8ed1ab_1\n",
            "Linking conda-package-handling-2.4.0-pyh7900ff3_2\n",
            "Linking conda-25.3.0-py312h7900ff3_0\n",
            "Linking conda-libmamba-solver-25.3.0-pyhd8ed1ab_0\n",
            "\n",
            "Transaction finished\n",
            "\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniforge3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniforge3: /root/miniforge3\n",
            "Do you wish to update your shell profile to automatically initialize conda?\n",
            "This will activate conda on startup and change the command prompt when activated.\n",
            "If you'd prefer that conda's base environment not be activated on startup,\n",
            "   run the following command when conda is activated:\n",
            "\n",
            "conda config --set auto_activate_base false\n",
            "\n",
            "You can undo this by running `conda init --reverse $SHELL`? [yes|no]\n",
            "[no] >>> yes\n",
            "no change     /root/miniforge3/condabin/conda\n",
            "no change     /root/miniforge3/bin/conda\n",
            "no change     /root/miniforge3/bin/conda-env\n",
            "no change     /root/miniforge3/bin/activate\n",
            "no change     /root/miniforge3/bin/deactivate\n",
            "no change     /root/miniforge3/etc/profile.d/conda.sh\n",
            "no change     /root/miniforge3/etc/fish/conf.d/conda.fish\n",
            "no change     /root/miniforge3/shell/condabin/Conda.psm1\n",
            "no change     /root/miniforge3/shell/condabin/conda-hook.ps1\n",
            "no change     /root/miniforge3/lib/python3.12/site-packages/xontrib/conda.xsh\n",
            "no change     /root/miniforge3/etc/profile.d/conda.csh\n",
            "modified      /root/.bashrc\n",
            "\n",
            "==> For changes to take effect, close and re-open your current shell. <==\n",
            "\n",
            "Running `shell init`, which:\n",
            " - modifies RC file: \"/root/.bashrc\"\n",
            " - generates config for root prefix: \u001b[1m\"/root/miniforge3\"\u001b[0m\n",
            " - sets mamba executable to: \u001b[1m\"/root/miniforge3/bin/mamba\"\u001b[0m\n",
            "The following has been added in your \"/root/.bashrc\" file\n",
            "\u001b[32m\n",
            "# >>> mamba initialize >>>\n",
            "# !! Contents within this block are managed by 'mamba shell init' !!\n",
            "export MAMBA_EXE='/root/miniforge3/bin/mamba';\n",
            "export MAMBA_ROOT_PREFIX='/root/miniforge3';\n",
            "__mamba_setup=\"$(\"$MAMBA_EXE\" shell hook --shell bash --root-prefix \"$MAMBA_ROOT_PREFIX\" 2> /dev/null)\"\n",
            "if [ $? -eq 0 ]; then\n",
            "    eval \"$__mamba_setup\"\n",
            "else\n",
            "    alias mamba=\"$MAMBA_EXE\"  # Fallback on help from mamba activate\n",
            "fi\n",
            "unset __mamba_setup\n",
            "# <<< mamba initialize <<<\n",
            "\u001b[0m\n",
            "Thank you for installing Miniforge3!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!export PATH=$PATH:/root/miniforge3/bin/conda\n",
        "!/root/miniforge3/bin/conda init bash\n",
        "\n",
        "##Bioconda adding source\n",
        "!/root/miniforge3/bin/conda config --add channels defaults\n",
        "!/root/miniforge3/bin/conda config --add channels bioconda\n",
        "!/root/miniforge3/bin/conda config --add channels conda-forge\n",
        "\n",
        "## Update conda after installation\n",
        "!/root/miniforge3/bin/conda update conda\n",
        "!/root/miniforge3/bin/mamba update mamba"
      ],
      "metadata": {
        "id": "TGz5M0ewCsXW",
        "outputId": "7085d1d3-7ee5-4e64-c571-23207c530288",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "no change     /root/miniforge3/condabin/conda\n",
            "no change     /root/miniforge3/bin/conda\n",
            "no change     /root/miniforge3/bin/conda-env\n",
            "no change     /root/miniforge3/bin/activate\n",
            "no change     /root/miniforge3/bin/deactivate\n",
            "no change     /root/miniforge3/etc/profile.d/conda.sh\n",
            "no change     /root/miniforge3/etc/fish/conf.d/conda.fish\n",
            "no change     /root/miniforge3/shell/condabin/Conda.psm1\n",
            "no change     /root/miniforge3/shell/condabin/conda-hook.ps1\n",
            "no change     /root/miniforge3/lib/python3.12/site-packages/xontrib/conda.xsh\n",
            "no change     /root/miniforge3/etc/profile.d/conda.csh\n",
            "no change     /root/.bashrc\n",
            "No action taken.\n",
            "/bin/bash: line 1: conda: command not found\n",
            "/bin/bash: line 1: conda: command not found\n",
            "/bin/bash: line 1: conda: command not found\n",
            "/bin/bash: line 1: conda: command not found\n",
            "/bin/bash: line 1: mamba: command not found\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Install Qiime2"
      ],
      "metadata": {
        "id": "r-sOEdtgCYle"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "fsSXV_IACXSu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# à¸ªà¹ˆà¸§à¸™à¹ƒà¸«à¸¡à¹ˆ"
      ],
      "metadata": {
        "id": "h0hva8vc_5aP"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nQRlv0ypvthh"
      },
      "source": [
        "# ðŸ¦  Amplicon Sequencing Data Analysis with Qiime 2\n",
        "\n",
        "This notebook will accompany the first session of the 2020 ISB Microbiome Course. The presentation slides can be [found here](https://gibbons-lab.github.io/isb_course_2020/16S).\n",
        "\n",
        "Save your own local copy of this notebook by using `File > Save a copy in Drive`. At some point you may be promted to certify the notebook is safe. We promise that it is ðŸ¤ž\n",
        "\n",
        "**Disclaimer:**\n",
        "\n",
        "The Google Colab notebook environment will interpret any command as Python code by default. If we want to run bash commands we will have to prefix them by `!`. So any command you see with a leading `!` is a bash command and if you wanted to run it in your terminal you would omit the leading `!`. For example, if in the Colab notebook you ran `!wget` you would just run `wget` in your terminal.\n",
        "\n",
        "## Setup\n",
        "\n",
        "QIIME 2 is usually installed by following the [official installation instructions](https://docs.qiime2.org/2021.4/install/). However, because we are using Google Colab and there are some caveats to using conda here, we will have to hack around the installation a little bit. But no worries, we provide a setup script below which does all this work for us. ðŸ˜Œ\n",
        "\n",
        "So...let's start by pulling a local copy of the project repository down from GitHub."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QFsZdvOuvZmq"
      },
      "source": [
        "!git clone https://github.com/gibbons-lab/isb_course_2020 materials"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fQVnDjFauZ-n"
      },
      "source": [
        "Now we are ready to set up our environment. This will take about 10-15 minutes.\n",
        "\n",
        "**Note**: This setup is only relevant for Google Colaboratory and will not work on your local machine. Please follow the [official installation instructions](https://docs.qiime2.org/2021.4/install/) for that."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BAyTazMCvrpf"
      },
      "source": [
        "%run materials/setup_qiime2.py"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2eenIEY76mcv"
      },
      "source": [
        "We will switch to working within the `materials` directory for the rest of the notebook."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "h3NALJ7u6mBP"
      },
      "source": [
        "%cd materials"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BGWsqUfq5oSi"
      },
      "source": [
        "## Our first QIIME 2 command\n",
        "\n",
        "Let's remember our workflow for today.\n",
        "\n",
        "![our workflow](https://github.com/Gibbons-Lab/isb_course_2020/raw/master/docs/16S/assets/steps.png)\n",
        "\n",
        "The first thing we have to do is to get the data into an artifact.\n",
        "We can import the data with the `import` action from the tools. For that we have to give\n",
        "QIIME 2 a *manifest* (list of raw files) and tell it what *type of data* we\n",
        "are importing and what *type of artifact* we want.\n",
        "\n",
        "**QoL Tip:** QIIME 2 commands can get very long. To split them up over several lines we can use `\\` which means \"continue on the next line\"."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6mGgUrRf5pyb"
      },
      "source": [
        "!qiime tools import \\\n",
        "  --type 'SampleData[SequencesWithQuality]' \\\n",
        "  --input-path manifest.tsv \\\n",
        "  --output-path cdiff.qza \\\n",
        "  --input-format SingleEndFastqManifestPhred33V2"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lw7TmzbiAoLj"
      },
      "source": [
        "Since we have quality information for the sequencing reads, let's also generate\n",
        "our first visualization to inspect sequence quality.\n",
        "\n",
        "---\n",
        "\n",
        "Qiime 2 commands can become pretty long. Here are some pointers to remember the\n",
        "structure of a command:\n",
        "\n",
        "```\n",
        "qiime plugin action --i-argument1 ... --o-argument2 ...\n",
        "```\n",
        "\n",
        "Argument types usually begin with a letter denoting their meaning:\n",
        "\n",
        "- `--i-...` = input files\n",
        "- `--o-...` = output files\n",
        "- `--p-...` = parameters\n",
        "- `--m-...` = metadata\n",
        "\n",
        "---\n",
        "\n",
        "In this case we will use the `summarize` action from the `demux` plugin with the previously generated artifact as input and output the resulting visualization to the `qualities.qzv` file."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "feGa41cEApGs"
      },
      "source": [
        "!qiime demux summarize --i-data cdiff.qza --o-visualization qualities.qzv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "O2GenFy3C8gu"
      },
      "source": [
        "You can view the plot by downloading the .qzv file and opening it using http://view.qiime2.org. To download the file click on the folder symbol to the left, open the `materials` folder, and choose download from the dot menu next to the `qualities.qzv` file.\n",
        "\n",
        "ðŸ¤” What do you observe across the read? Where would you truncate the reads?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "os5MeRvLIq0v"
      },
      "source": [
        "# Denoising amplicon sequence variants\n",
        "\n",
        "We will now run the DADA2 plugin which will do 3 things:\n",
        "\n",
        "1. filter and trim the reads\n",
        "2. find the most likely set of unique sequences in the sample (ASVs)\n",
        "3. remove chimeras\n",
        "4. count the abundances of each ASV\n",
        "\n",
        "\n",
        "Since this step takes a bit, let's start the process and use the time to\n",
        "understand what is happening:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RdLSMp9-Dl-d"
      },
      "source": [
        "!qiime dada2 denoise-single \\\n",
        "    --i-demultiplexed-seqs cdiff.qza \\\n",
        "    --p-trunc-len 150 \\\n",
        "    --output-dir dada2 --verbose"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aFpQQ5GHdbzX"
      },
      "source": [
        "If this step takes too long or fails, you can also copy the results from the treasure chest. *However, don't run the next cell if this cell completes successfully.*"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PFrfo-uCdoRz"
      },
      "source": [
        "# only run if the previous cell did not work\n",
        "!cp -r treasure_chest/dada2 ."
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4YzrovxVj7U_"
      },
      "source": [
        "Ok, this step ran, but we should also make sure it kind of worked. One good way to tell if the identified ASVs are representative of the sample is to see how many reads were maintained throughout the pipeline. Here, the most common issues and solutions are:\n",
        "\n",
        "**Large fraction of reads is lost during merging (only paired-end)**\n",
        "\n",
        "![read overlap](https://gibbons-lab.github.io/isb_course_2020/16S/assets/read_overlap.png)\n",
        "\n",
        "In order to merge ASVs DADA2 uses an overlap of 12 bases between forward and reverse reads by default. Thus, your reads must allow for sufficient overlap *after* trimming. So if your amplified region is 450bp long and you have 2x250bp reads and you trim the last 30 bases of each read, truncating the length to 220bp, the total length of covered sequence is 2x220 = 440 which is shorter than 450bp so there will be no overlap. To solve this issue trim less of the reads or adjust the `--p-min-overlap` parameters to something lower (but not too low).\n",
        "\n",
        "<br>\n",
        "\n",
        "**Most of the reads are lost as chimeric**\n",
        "\n",
        "![read overlap](https://gibbons-lab.github.io/isb_course_2020/16S/assets/chimera.png)\n",
        "\n",
        "This is usualluy an experimental issue as chimeras are introduced during amplification. If you can adjust your PCR, try to run fewer cycles. Chimeras can also be introduced by incorrect merging. If your minimum overlap is too small ASVs may be merged randomly. Possible fixes are to increase the `--p-min-overlap` parameter or run the analysis on the forward reads only (in our empirical observations, chimeras are more likely to be introduced in the joined reads). *However, losing between 5-25% of your reads to chimeras is normal and does not require any adjustments.*\n",
        "\n",
        "Our denoising stats are contained in an artifact. To convert it to a visualization we can use `qiime metadata tabulate`."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "n_Tr3Ujcj61w"
      },
      "source": [
        "!qiime metadata tabulate \\\n",
        "    --m-input-file dada2/denoising_stats.qza \\\n",
        "    --o-visualization dada2/denoising-stats.qzv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "o75g7DSO2tu5"
      },
      "source": [
        "What proprotion of reads was retained throughout the entire pipeline? Look at the final number of used reads (non-chimeric). What do you observe when comparing those values between samples and how might that affect diversity metrics?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "c0ZV3_sGQ20u"
      },
      "source": [
        "# Phylogeny and diversity"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aEit8p7V9Ht9"
      },
      "source": [
        "## Building a tree\n",
        "\n",
        "We can build a phylogenetic tree for our sequences using the following command:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "2PVNlD0g9MsX"
      },
      "source": [
        "!qiime phylogeny align-to-tree-mafft-fasttree \\\n",
        "    --i-sequences dada2/representative_sequences.qza \\\n",
        "    --output-dir tree"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UmxTRou99a_n"
      },
      "source": [
        "You can visualize your tree using iTOL (https://itol.embl.de/). Open iTol and upload the artifact from `materials/tree/tree.qza`.\n",
        "\n",
        "This looks nice but is not particularly informative as is. The main utility of our tree will be in complementing our diversity analyses. It will tell us which ASVs are more or less related to one another, which will allow us to calculate different kinds of ecological diversity metrics.\n",
        "\n",
        "## Alpha and Beta Diversity\n",
        "\n",
        "Qiime 2 has \"run-all\" command for diversity analyses. This will\n",
        "\n",
        "1. Subsampe our samples to the same total number of reads (Why?)\n",
        "2. Calculate alpha and beta diversity measures\n",
        "3. Visualize PCoA projections"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0js6xHuw-NqF"
      },
      "source": [
        "!qiime diversity core-metrics-phylogenetic \\\n",
        "    --i-table dada2/table.qza \\\n",
        "    --i-phylogeny tree/rooted_tree.qza \\\n",
        "    --p-sampling-depth 8000 \\\n",
        "    --m-metadata-file metadata.tsv \\\n",
        "    --output-dir diversity"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "p-uZBNVX-y2L"
      },
      "source": [
        "## Statistical analyses\n",
        "\n",
        "Let's first have a look at alpha diversity. Can we see a difference in the per-sample diversity between healthy and sick individuals?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GnkeX6iY--g-"
      },
      "source": [
        "!qiime diversity alpha-group-significance \\\n",
        "    --i-alpha-diversity diversity/shannon_vector.qza \\\n",
        "    --m-metadata-file metadata.tsv \\\n",
        "    --o-visualization diversity/alpha_groups.qzv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ngH5fUeDQMO0"
      },
      "source": [
        "Now, let's use beta diversity to see how different the samples are from one another. First download `diversity/weighted_unifrac_emperor.qzv` and take a look. Do samples separate based on disease state?\n",
        "\n",
        "We can check whether that separation is 'significant' by using a PERMANOVA test."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KzNW92riQguX"
      },
      "source": [
        "!qiime diversity adonis \\\n",
        "    --i-distance-matrix diversity/weighted_unifrac_distance_matrix.qza \\\n",
        "    --m-metadata-file metadata.tsv \\\n",
        "    --p-formula \"disease_stat\" \\\n",
        "    --p-n-jobs 2 \\\n",
        "    --o-visualization permanova.qzv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FB7WYPZ1DEoR"
      },
      "source": [
        "# Taxonomy\n",
        "\n",
        "We will use a Bayes classifier trained on the GreenGenes database which can be downloaded from https://docs.qiime2.org/2021.4/data-resources/."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "drgYtfLbDosv"
      },
      "source": [
        "!wget https://data.qiime2.org/2021.4/common/gg-13-8-99-515-806-nb-classifier.qza"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EEBbgvnhDGTs"
      },
      "source": [
        "!qiime feature-classifier classify-sklearn \\\n",
        "    --i-reads dada2/representative_sequences.qza \\\n",
        "    --i-classifier gg-13-8-99-515-806-nb-classifier.qza \\\n",
        "    --o-classification taxa.qza"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xISxzBo6Mpka"
      },
      "source": [
        "Now let's have a look at the relative abundances of the different bacterial taxa we have in each sample:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4VTL2yEBMqfO"
      },
      "source": [
        "!qiime taxa barplot \\\n",
        "    --i-table dada2/table.qza \\\n",
        "    --i-taxonomy taxa.qza \\\n",
        "    --m-metadata-file metadata.tsv \\\n",
        "    --o-visualization taxa_barplot.qzv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4_FVM8VfX_yQ"
      },
      "source": [
        "We can also collapse data on a particular taxonomic rank using the QIIME 2 [taxa plugin](https://docs.qiime2.org/2021.4/plugins/available/taxa/). Why might we want to look at different taxonomic ranks, rather than just looking at ASVs?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gl72G8KaYmBm"
      },
      "source": [
        "!qiime taxa collapse \\\n",
        "    --i-table dada2/table.qza \\\n",
        "    --i-taxonomy taxa.qza \\\n",
        "    --p-level 6 \\\n",
        "    --o-collapsed-table genus.qza"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dlVdzgfbZcjt"
      },
      "source": [
        "We can export the table and convert it to a .csv file so that we can analyze these data using tools outside of the QIIME 2 environment."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZsxKf3yAZiHk"
      },
      "source": [
        "!qiime tools export \\\n",
        "    --input-path genus.qza \\\n",
        "    --output-path exported\n",
        "!biom convert -i exported/feature-table.biom -o genus.tsv --to-tsv"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TezGLSEGZrcn"
      },
      "source": [
        "Now the data are in a common format and we can use them, for instance, to draw a heatmap using Pandas and Seaborn."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "p5zUjU8JZqL4"
      },
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "\n",
        "abundances = pd.read_table(\"genus.tsv\", skiprows=1, index_col=0)\n",
        "abundances.index = abundances.index.str.split(\";\").str[5]  # Use only the genus name\n",
        "abundances = abundances[~abundances.index.isin([\"g__\", \"__\"])]  # remove unclassified genera\n",
        "\n",
        "# Let's do a centered log-ratio transform: log x_i - log mean(x)\n",
        "transformed = abundances.apply(\n",
        "    lambda xs: np.log(xs + 0.5) - np.log(xs.mean() + 0.5),\n",
        "    axis=1)\n",
        "\n",
        "sns.clustermap(transformed.T, cmap=\"magma\", xticklabels=True, figsize=(16, 5))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QbRoCh1BR0GZ"
      },
      "source": [
        "# Exercises\n",
        "\n",
        "Okay, that's enough time in the back seat.\n",
        "\n",
        "It's time to take the wheel ðŸš—\n",
        "\n",
        "Now you can dive into the data ðŸŠ"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Moq2YOvDSDap"
      },
      "source": [
        "## Exercise 1 - Sample Classification\n",
        "\n",
        "One pretty basic question we can ask is whether the microbial composition in the gut is predictive of recurrent *C. difficile* infection.\n",
        "Could you predict if someone had a recurrent infection from 16S data alone?\n",
        "\n",
        "Metadata about the samples is again contained in `metadata.tsv` which looks like this:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "T16jC6U1ONow"
      },
      "source": [
        "import pandas as pd\n",
        "\n",
        "pd.read_csv(\"metadata.tsv\", sep=\"\\t\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "j1Q2ZpBEON42"
      },
      "source": [
        "Start with the `classify-samples` action and follow it up by finding and looking at the `heatmap` visualization afterwards that shows important genera. You can also find accuracy results that will tell you how well the phenotype can be predicted from the microbial composition."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Nhf1g2RpSCak"
      },
      "source": [
        "# Your code here"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hy5HN4ZoS2-9"
      },
      "source": [
        "## Exercise 2 - Plant a Tree\n",
        "\n",
        "One visualization that we did not spend a lot of time on was the phylogentic tree of our ASVs. Let's change that!\n",
        "\n",
        "First, annotate the tree with taxonomy by going back to the tree (uploading `tree.qza`) and visualizing it using iTOL (https://itol.embl.de/). You can add the predicted taxa (`taxa.qza`) using the `+` button on the lower right of the visualization.\n",
        "\n",
        "Also add your feature table, but don't use the raw one (why?). Rather, use `diversity/rarefied_table.qza` where each sample has been subsampled to 10000 reads (why?).\n",
        "\n",
        "What other normalization methods could you use?\n",
        "\n",
        "What do you observe?"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bx6qkBT-85VE"
      },
      "source": [],
      "execution_count": null,
      "outputs": []
    }
  ]
}